python/tank/bootstrap/async.py

```diff
@@ -0,0 +1,278 @@
+        """
+
+        super(AsyncBootstrapWrapper, self).__init__()
+
+        self._toolkit_manager = toolkit_manager
+        self._engine_name = engine_name
+        self._entity = entity
+
+        # Retrieve the toolkit manager progress reporting callback.
+        self._progress_callback = self._toolkit_manager._progress_cb
+
+        self._completed_callback = None
+        self._failed_callback = None
+
+        # Create a worker that can bootstrap the toolkit asynchronously in a background thread.
+        self._worker = _BootstrapToolkitWorker(self._toolkit_manager, engine_name, entity)
+
+        # This QThread object will live in the main thread, not in the new thread it will manage.
+        self._thread = QtCore.QThread()
+
+        # Make the worker operate with the new thread affinity and use the QThread object event loop.
+        self._worker.moveToThread(self._thread)
+
+        # Start to work when the QThread object will have started its event loop in its new thread context.
+        self._thread.started.connect(self._worker.work)
+
+        # Make the worker report on the toolkit bootstrap progress.
+        self._worker.progressing.connect(self._progress_bootstrap)
+
+        # Handle completion of the toolkit bootstrap by the worker.
+        self._worker.completed.connect(self._complete_bootstrap)
+
+        # Handle failure of the toolkit bootstrap by the worker.
+        self._worker.failed.connect(self._fail_bootstrap)
+
+    def set_callbacks(self, completed_callback, failed_callback):
+        """
+        Sets the callback functions used to signify the completion and the failure of the bootstrap.
+
+        These functions must have the following signatures:
+
+            ``completed_callback(engine)``
+
+            where:
+            - ``engine`` is the launched :class:`~sgtk.platform.Engine` instance.
+
+            ``failed_callback(phase, exception)``
+
+            where:
+            - ``phase`` is the bootstrap phase that raised the exception,
+                        ``ToolkitManager.TOOLKIT_BOOTSTRAP_PHASE`` or ``ToolkitManager.ENGINE_STARTUP_PHASE``.
+                        Using this phase, the callback can decide if the toolkit core needs
+                        to be re-imported to ensure usage of a swapped in version.
+            - ``exception`` is the python exception raised while bootstrapping.
+
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        """
+
+        self._completed_callback = completed_callback
+        self._failed_callback = failed_callback
+
+    def bootstrap(self):
+        """
+        Starts the asynchronous bootstrap logic.
+        """
+
+        # Start the QThread object event loop in its new thread context.
+        self._thread.start()
+
+    @QtCore.Slot(float, str, int, int)
```

Comment:
So sorry!! I totally missed those comments due to the collapsing in github! 
I would suggest adding a comment in this case to highlight the special case:

```python
# note: we are using the slot decorators because of threading
# http://stackoverflow.com/questions/14421897/is-the-pyside-slot-decorator-necessary
@QtCore.Slot(float, str, int, int)
def _progress_bootstrap(self, progress_value, message, current_index, maximum_index):
```